research/train.py

```diff
@@ -37,20 +44,59 @@ def _update(engine, batch):
         loss = loss_fn(y_pred, y)
         loss.backward()
         optimizer.step()
-        return loss.item(), y_pred, y
 
-    def _metrics_transform(output):
-        return output[1], output[2]
+        # Anything you want to log must be returned in this dictionary
+        update_dict = {"nll": loss.item(), "y_pred": y_pred, "y": y}
+        return update_dict
 
     engine = Engine(_update)
 
-    for name, metric in metrics.items():
-        metric._output_transform = _metrics_transform
-        metric.attach(engine, name)
+    # Common metrics require base "y_pred" and "y". Lambda to reduce verbosity in metrics
+    _ypred_y = lambda _: (_["y_pred"], _["y"])
+
+    # (Optional) Specify training metrics. "output_transform" used to select items from "update_dict" needed by metrics
+    # Collecting metrics over training set is not recommended
+    # https://pytorch.org/ignite/metrics.html#ignite.metrics.Loss
 
     return engine
 
 
+def create_supervised_evaluator(
+    model: nn.Module, cfg: DictConfig, device="cpu"
+) -> Engine:
+
+    # Loss
+    loss_fn = torch.nn.NLLLoss()
+
+    def _inference(engine, batch):
+        model.eval()
+        with torch.no_grad():
+            x, y = batch
+            x, y = x.to(device), y.to(device)
+            y_pred = model(x)
+
+        # Anything you want to log must be returned in this dictionary
+        infer_dict = {"y_pred": y_pred, "y": y}
+        return infer_dict
+
+    evaluator = Engine(_inference)
+
+    # Common metrics require base "y_pred" and "y". Lambda to reduce verbosity in metrics
+    _ypred_y = lambda _: (_["y_pred"], _["y"])
+
+    # Specify evaluation metrics. "output_transform" used to select items from "infer_dict" needed by metrics
+    # https://pytorch.org/ignite/metrics.html#ignite.metrics.Loss
+    metrics = {
+        "accuracy": Accuracy(output_transform=_ypred_y),
+        "nll": Loss(loss_fn, output_transform=_ypred_y),
```

Comment:
The final loss and accuracy is done over the entire eval dataset correct?

Reply:
Correct, for this demo! But we could accumulate loss/acc over the training set as we iterate if we wanted. Just depends on the major use case